xrefs extraction tool

eqvinox opened the pull request:
This is `xrelfo.py`, the tool to extract the xrefs added in PR #6766 

I'm still polishing up the Python code but it's already quite workable and ready to be looked at.

riw777 reviewed (approved):
looks good -- let me know when you're ready for a final look and commit

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/8be0ef2778de1c74d7bc6a4ac83a24e5/raw/b5a788d2ca9c8c65b30f93b744a0c1b648fe59a0/cr_6807_1613268051.diff | git apply
```

</p>

<p>

```diff
diff --git a/lib/command_graph.h b/lib/command_graph.h
index 86715410c..f3c3dd43d 100644
--- a/lib/command_graph.h
+++ b/lib/command_graph.h
@@ -99,7 +99,7 @@ struct cmd_element {
 	const char *string; /* Command specification by string. */
 	const char *doc;    /* Documentation of this command. */
 	int daemon;	 /* Daemon to which this command belong. */
-	uint32_t attr;       /* Command attributes */
+	uint32_t attr;   /* Command attributes */
 
 	/* handler function for command */
 	int (*func)(const struct cmd_element *, struct vty *, int,
diff --git a/lib/zlog.h b/lib/zlog.h
index 2aaa2ebcf..4c5e3d7a4 100644
--- a/lib/zlog.h
+++ b/lib/zlog.h
@@ -85,19 +85,21 @@ static inline void zlog_ref(const struct xref_logmsg *xref,
 	va_end(ap);
 }
 
-#define _zlog_ref(prio, msg, ...) do {                                         \
+#define _zlog_ref(prio, msg, ...)                                              \
+	do {                                                                   \
 		static struct xrefdata _xrefdata = {                           \
 			.hashstr = (msg),                                      \
-			.hashu32 = { (prio), 0 },                              \
+			.hashu32 = {(prio), 0},                                \
 		};                                                             \
-		static const struct xref_logmsg _xref __attribute__((used)) = {\
+		static const struct xref_logmsg _xref __attribute__(           \
+			(used)) = {                                            \
 			.xref = XREF_INIT(XREFT_LOGMSG, &_xrefdata, __func__), \
 			.fmtstring = (msg),                                    \
 			.priority = (prio),                                    \
 			.args = (#__VA_ARGS__),                                \
 		};                                                             \
 		XREF_LINK(_xref.xref);                                         \
-		zlog_ref(&_xref, (msg), ## __VA_ARGS__);                       \
+		zlog_ref(&_xref, (msg), ##__VA_ARGS__);                        \
 	} while (0)
 
 #define zlog_err(...)    _zlog_ref(LOG_ERR, __VA_ARGS__)
@@ -106,12 +108,14 @@ static inline void zlog_ref(const struct xref_logmsg *xref,
 #define zlog_notice(...) _zlog_ref(LOG_NOTICE, __VA_ARGS__)
 #define zlog_debug(...)  _zlog_ref(LOG_DEBUG, __VA_ARGS__)
 
-#define _zlog_ecref(ec_, prio, msg, ...) do {                                  \
+#define _zlog_ecref(ec_, prio, msg, ...)                                       \
+	do {                                                                   \
 		static struct xrefdata _xrefdata = {                           \
 			.hashstr = (msg),                                      \
-			.hashu32 = { (prio), (ec_) },                          \
+			.hashu32 = {(prio), (ec_)},                            \
 		};                                                             \
-		static const struct xref_logmsg _xref __attribute__((used)) = {\
+		static const struct xref_logmsg _xref __attribute__(           \
+			(used)) = {                                            \
 			.xref = XREF_INIT(XREFT_LOGMSG, &_xrefdata, __func__), \
 			.fmtstring = (msg),                                    \
 			.priority = (prio),                                    \
@@ -119,7 +123,7 @@ static inline void zlog_ref(const struct xref_logmsg *xref,
 			.args = (#__VA_ARGS__),                                \
 		};                                                             \
 		XREF_LINK(_xref.xref);                                         \
-		zlog_ref(&_xref, "[EC %u] " msg, ec_, ## __VA_ARGS__);         \
+		zlog_ref(&_xref, "[EC %u] " msg, ec_, ##__VA_ARGS__);          \
 	} while (0)
 
 #define flog_err(ferr_id, format, ...)                                         \
diff --git a/python/clippy/__init__.py b/python/clippy/__init__.py
index 344a1c91e..dac8eb4b1 100644
--- a/python/clippy/__init__.py
+++ b/python/clippy/__init__.py
@@ -21,7 +21,10 @@ import _clippy
 from _clippy import parse, Graph, GraphNode
 
 
-frr_top_src = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+frr_top_src = os.path.dirname(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+)
+
 
 def graph_iterate(graph):
     """iterator yielding all nodes of a graph
diff --git a/python/makefile.py b/python/makefile.py
index 44658013b..e78aa1f4e 100644
--- a/python/makefile.py
+++ b/python/makefile.py
@@ -32,7 +32,12 @@ for clippy_file in clippy_scan:
     assert clippy_file.endswith(".c")
 
 xref_targets = []
-for varname in ["bin_PROGRAMS", "sbin_PROGRAMS", "lib_LTLIBRARIES", "module_LTLIBRARIES"]:
+for varname in [
+    "bin_PROGRAMS",
+    "sbin_PROGRAMS",
+    "lib_LTLIBRARIES",
+    "module_LTLIBRARIES",
+]:
     xref_targets.extend(mv[varname].strip().split())
 
 # check for files using clippy but not listed in clippy_scan
@@ -129,12 +134,14 @@ for clippy_file in clippy_scan:
     out_lines.append(clippydep.substitute(clippybase=clippy_file[:-2]))
 
 out_lines.append("")
-out_lines.append("xrefs = %s" % (" ".join(["%s.xref" % target for target in xref_targets])))
+out_lines.append(
+    "xrefs = %s" % (" ".join(["%s.xref" % target for target in xref_targets]))
+)
 out_lines.append("frr.xref: $(xrefs)")
 out_lines.append("")
 
-#frr.xref: $(bin_PROGRAMS) $(sbin_PROGRAMS) $(lib_LTLIBRARIES) $(module_LTLIBRARIES)
-#	$(AM_V_XRELFO) $(CLIPPY) $(top_srcdir)/python/xrelfo.py -o $@ $^
+# frr.xref: $(bin_PROGRAMS) $(sbin_PROGRAMS) $(lib_LTLIBRARIES) $(module_LTLIBRARIES)
+# 	$(AM_V_XRELFO) $(CLIPPY) $(top_srcdir)/python/xrelfo.py -o $@ $^
 
 out_lines.append("")
 out_lines.extend(bcdeps)

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

eqvinox commented:
@louberger need logs.  Most likely issue is that `libelf-dev` is now required to build, please install if needed.

eqvinox commented:
(NB: `libelf-dev` is a build time dependency only, runtime dependencies have not changed.)

eqvinox commented:
@riw777 @Jafaral ready! Changes since you looked at it:

* `frr.xref` is now built by default on all systems during `make`
* `frr.xref` is now built in parallel for each binary & combined at the end
* CI system now saves the `frr.xref` file as an artifact on test runs (compressed with `xz`)
* for log messages, the format arguments are now recorded too.  This allows `xrelfo.py` to "suggest" things like...
* new warnings/"suggestions" from `xrelfo.py`: (these are not build errors)
  * `cleanup: replace inet_ntop(AF_INET, ...) with %pI4`
  * `cleanup: replace inet_ntop(AF_INET6, ...) with %pI6`
  * `cleanup: replace inet_ntoa(...) with %pI4`
  * `cleanup: replace ipaddr2str(...) with %pIA`
  * `cleanup: replace prefix2str(...) with %pFX`
  * `cleanup: replace prefix_mac2str(...) with %pEA`
  * `cleanup: replace sockunion2str(...) with %pSU`
* previously existing warnings are now printed during `make`: (again, not build errors)
  * `error: log message contains tab or newline`
  * `warning: log message starts with severity`
* (details for warnings in https://github.com/FRRouting/frr/pull/6807/files#diff-f482f7cf6380cc0724a64dd1fd0c46fa16673b24f25bb6266559e405efe9355cR118-R134 )
  * the warnings are printed only if `--enable-dev-build` was given on `configure`; there's quite a few of them and they're colored a nice red :red_square: to give everyone an incentive to look at some of them
* added `binary` field to output JSON which lists the binary the item was extracted from, to avoid guessing based on source filename
* removed Python handling for padding in C structs.  Only `struct cmd_element` had 3 bytes of padding after `uint8_t attr`, which I replaced with `uint32_t` (no padding) for the sake of simplicity.  (Handling padding correctly would have significantly complicated the Python code for 32/64bit cases & since we're controlling the entire situation here why make it harder than it needs be?)
* random fixes to get to 100% on all CI systems (e.g. OpenBSD ELF is slightly weird :laughing:)
* fixed style warnings

riw777 reviewed (approved):
looks good ... will wait a bit to see if anyone else wants to review before pushing

eqvinox commented:
@riw777 noone seems to care enough to say anything :)

qlyoung commented:
If only I had unlimited time to review every PR that's been unloved for so long

qlyoung reviewed (dismissed):
Frankly I have no idea what this does from skimming the code, can you add a section on how to use this thing to the xrefs doc?

As an aside I think it's probably fair to say nobody is going to put meaningful time into reviewing ELF parsing code so once there's a section in the doc explaining what this actually does lets just merge it

eqvinox commented:
> Frankly I have no idea what this does from skimming the code, can you add a section on how to use this thing to the xrefs doc?

Will do tomorrow.

eqvinox commented:
@qlyoung docs added.

eqvinox commented:
Missed a style warning, fixed.  The "ELFFormatError" one is bogus, probably checkpatch.pl thinks ELFFormatError is a macro because it starts with an uppercase letter.  But it's a variable in Python naming scheme.

louberger reviewed (dismissed):
the build docs need to be updated to note that libelf-dev (or equivalent) is needed.  Following the install current directions i get:
configure: error: libelf headers are required for building clippy.  (Host only when cross-compiling.)

eqvinox commented:
@louberger done, thanks, easy to forget these.

(fun fact: libelf is a system component on the BSDs [and Solaris, not that we care])

louberger commented:
Yup, this is why we have automated testing...

________________________________

On February 24, 2021 1:18:19 AM David Lamparter <notifications@github.com> wrote:

@louberger<https://github.com/louberger> done, thanks, easy to forget these.

(fun fact: libelf is a system component on the BSDs [and Solaris, not that we care])

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/FRRouting/frr/pull/6807#issuecomment-784818765>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABSFNSBUZ3Z63AHPU3YBO53TASK2RANCNFSM4PI7VNVA>.

eqvinox commented:
anyone feel like pressing the merge button? @riw777 @Jafaral ?

Much appreciated in advance :relaxed:

eqvinox commented:
weekly keepalive and note that I'm unaware of anything left to do here

riw777 reviewed (approved):
looks good ... I think everything has been covered.
